### DISCUSSION TITLE
CSL, is it the right choice for typist?

### DISCUSSION THREAD

**Discussion by DrGo** [typst/typst#2406] [Ideas]
https://docs.citationstyles.org/en/stable/specification.html
I read somewhere that Typist is standardizing on using CSL styles for formatting citations and bibliographies. If true I think this is a step backwards. The way I see it is that CSL is an awkward solution that uses a data exchange format, xml, to solve a problem (formatting text) that requires a simple, flexible and dynamic language (which by the way already exists: Typist !!).

I prefer to use Typist data structures and proper programming language grammar (functions, if statements etc) to using CSL’s verbose, complicated and limited syntax. XML was never meant to express complex logic and it shows e.g., in CSL awkward choose and if elements.

The only advantage of using CSL is that it has a large collection of styles; largely small variations on 3-4 themes. But it would be much simpler and probably more efficient and easier to maintain and document to transpile those into typist templates using a simple utility. That way you won’t have to be limited to what CSL can do and you won’t have to bloat Typist interface  and its ecosystem with CSL. (Formatting biblios will be just 
Typist code). Also, the same approach can be used to accommodate the equally large and equally awkward and limited RM output styles. 

Has this idea been considered?

**Comment by Dherse** [comment]
Actually you might have hit this right on the head, perhaps a code generator that takes in csl files and translates them to typst code that takes for input a pre-processed bib file would be excellent and even more flexible @reknih 

Maybe I am dead wrong, but it does sound like a nice alternative

**Comment by andreasKroepelin** [comment]
Obligatory "it is called Typst, without the i" comment.

Contentwise I can agree, maybe a modern alternative (analogously to BibTeX -> hayagriva) would be a good idea.

**Reply by DrGo** [reply]
Sorry about the typo.. funny I wrote first as Typst but went back to "fix" it .hahaha

**Reply by sitandr** [reply]
I just want to put it here (from the main Github page of Typst):

# Pronunciation and Spelling

IPA: /taɪpst/. "Ty" like in Typesetting and "pst" like in Hipster. When writing about Typst, capitalize its name as a proper noun, with a capital "T".

**Comment by Enivex** [comment]
How is it a step backwards? Customizing bibliographies is basically impossible right now.

Keep in mind that the vast, vast majority of users aren't going to be interested in adjusting the minute details of how the bibliography is presented. They just want to download the style file for their field or journal and be done with it. By doing with CSL we get thousands (?) of pre-existing templates for free.

This doesn't preclude exploring other options in the future.

**Reply by DrGo** [reply]
They could still do that or even use other style output formats if they want... the difference is that CSL or RM styles won't become enshrined as part of Typst interface. My proposal is that CSL or a similar format get transpiled into straightforward Typst script, so the compiler proper (and its maintainers) do not have to worry about CSL compatibility. Typst would be then free to have its own flexible and user-friendly interface which hopefully will be adopted by people who do not enjoy modifying CSL/RM styles. Meanwhile, all these styles are available for Typst users who just want to use an existing style of any kind (all is needed is an independent transpiler).

**Reply by DrGo** [reply]
It's a step backwards because I would like Typst to stay small and easy to learn (and maintain). The bigger the interface, the more difficult to learn and maintain, and the more likely that it will interact with other features to generate growing complexity (complexity is multiplicative as they say). Once you support CSL, not only that you will need to support it in perpetuity, you will also be expected to support other style formats or the next format that is going to replace CSL (there will be one!!).

**Reply by DrGo** [reply]
On the short-term, I imagine that Typst will bundle a small utility to do the transpilation. But in the future, Typst users could just use pre-transpiled styles downloadable from wherever they are hosted.

**Comment by DrGo** [comment]
A little bit more concrete proposal: 
A CSL is essentially a collection of static structs and macros. 
A transpiler will convert the structs (e.g., the Root.Info entity) into a Typst functions that returns a dict. Formatting elements (including macros) are translated into Typst fn that accept a bibliographic record as input and return another fn that specifies the desired formatting. The whole thing is stored as a module that exposes a set of fns that correspond to each element of the CSL style. The formatter loads the module and calls these fns to figure out what needs to be done; not much different from the current approach but instead of #text or #rect, you will have #name, #date etc. These in turn become target for transpilation from other formats or for writing Typst styles by hand.

**Comment by PgBiel** [comment]
I think it is legitimate to request for a pure-typst way to customize bibliography, as has been requested before (perhaps using functions taking some info dict and producing content output).

However, I also think it's perfectly fine to have the CSL option. It's additive - if you have a CSL style ready, well, now you will be able to use it; otherwise, you just remain with the status quo. In the future, some dict -> content function option for custom bib styles _wouldn't_ negate CSL styles, and both could coexist just fine. So I don't think it's a _problem_ (or "step backwards") to support CSL - rather, it'll just enable more users to be able to use typst for their workflow (which _usually_ involves some established bibliography style).

(I also don't believe CSL would become "enshrined" within the Typst interface, as I'm assuming the Typst interface will mostly just correspond to the selection of the desired CSL spec by the end user.)

**Reply by DrGo** [reply]
If CSL is natively supported it becomes part of the interface of Typist and you would not be able to change without breaking existing plugins and code.

**Reply by goyalyashpal** [reply]
STOP calling it with "i" please?

**Comment by emilyyyylime** [comment]
I'll mention I know close to nothing about citations, bibliographies, and CSL, but I agree with this deeply. Being able to transpile CSL into Typst or even dynamically interpret it with builtin functions is a much cleaner approach than tacking on another file format that Typst supports (or requires, depending on your use).

**Comment by reknih** [comment]
It is true that CSL is oftentimes awkward to process because it ultimately needs to introduce many different but similar domain-specific options, lacking the capabilities of a full programming language. It also does not handle bibliographies with entries with differing scripts/locales or legal citations well (neither does Hayagriva in its current state; also see CSL-M).

However, the CSL implementation will unlock not only 2,000 citation styles but also localization of terms in the references. I have personally invested over 2 weeks into implementing a CSL parser and processor. This implementation is about 85% done. We believe that this implementation will be a huge benefit for our users and cannot justify reimplementing with a slightly different approach time-wise.

We do, however, plan to expose the Hayagriva data structures in Typst so package authors can choose to implement a style there. A lot of new code for formatting various data types has been written for CSL, over time it could be factored out of CSL processing and provided in Typst to make this process easier.

**Comment by laurmaedje** [comment]
I completely agree with you that a simple, flexible and dynamic language is a nice way to format a bibliography. The problem is that there is no collection of citation styles written in such a language (be it Typst or another one). 

In theory, transpiling CSL to Typst sounds nice (and I have actually considered it). In practice, however, it would be a nightmare. CSL is hugely complex. Generating spec-compliant Typst code that formats like a specific CSL style would likely mean generating thousands of lines of Typst code handling case folding, localization, specific edge cases etc. The transpiler would be far more complex than the current CSL processor and the generated Typst code would probably be slow and barely editable by hand. I see no benefit to be had there.

That said, we do also plan to expose hayagriva's data model as Typst dictionaries alongside (or soon after) the CSL update. That way, everybody is free to write custom/new citation styles in Typst.

**Reply by bdarcus** [reply]
CSL creator here; interesting thread, with fair critiques.

> In theory, transpiling CSL to Typst sounds nice (and I have actually considered it). In practice, however, it would be a nightmare. CSL is hugely complex.

True. 

Edit: well, to be more specific, the template part of the language is complex.

I've long wondered about the possibility of using ML to derive styles from formatted output (the [another option](https://github.com/bdarcus/csln/issues/45) here), but that may just be a dream.

Worth noting that I, not a trained programmer, created the first prototype almost 20 years ago (~2005).

There are some limitations that come with that, in that I think some of it (specifically the template part) in retrospect is too complex, and other aspects too simple. But this was the first attempt to do this, and we weren't really sure what we needed at that point.

The core requirement of a language-agnostic style format, which is its primary advantage compared to all other options, also provides constraints, already discussed above.

This is my second attempt at the basic idea, still experimental at this point. My hope is it retains the 1.0 advantages and reduces some of its limitations.

https://github.com/bdarcus/csln

**Comment by DrGo** [comment]
Sounds good ) good luck with all that
